web3/packages/api-server/src/cache/guard.ts

```diff
@@ -55,79 +60,75 @@ export class AccessGuard {
     this.rpcMethods = config.methods;
     this.expiredTimeMilsecs = expiredTimeMilsecs || CACHE_EXPIRED_TIME_MILSECS;
     this.batchLimit = config.batch_limit || BATCH_LIMIT;
-  }
-
-  async setMaxReqLimit(rpcMethod: string, maxReqCount: number) {
-    this.rpcMethods[rpcMethod] = maxReqCount;
-  }
-
-  async getCount(rpcMethod: string, reqId: string) {
-    const id = getId(rpcMethod, reqId);
-    const count = await this.store.get(id);
-    if (count == null) {
-      return null;
-    }
-    return +count;
-  }
 
-  async add(rpcMethod: string, reqId: string): Promise<HexString | undefined> {
-    const isExist = await this.isExist(rpcMethod, reqId);
-    if (!isExist) {
-      const id = getId(rpcMethod, reqId);
-      await this.store.insert(id, 0);
-      return id;
-    }
+    this.store.scriptLoad(RATE_LIMIT_SCRIPT).then((evalSha) => {
+      this.evalSha = evalSha;
+      logger.info("AccessGuard eval sha:", evalSha);
+    });
   }
 
-  async updateCount(rpcMethod: string, reqId: string, offset: number = 1) {
-    const isExist = await this.isExist(rpcMethod, reqId);
-    if (isExist === true) {
-      const id = getId(rpcMethod, reqId);
-      if (offset > 1) {
-        await this.store.incrBy(id, offset);
-      } else {
-        await this.store.incr(id);
-      }
+  private async getEvalSha() {
+    if (!this.evalSha) {
+      this.evalSha = await this.store.scriptLoad(RATE_LIMIT_SCRIPT);
+      logger.warn(`AccessGuard reget eval sha:`, this.evalSha);
     }
+    return this.evalSha;
   }
 
-  async isExist(rpcMethod: string, reqId: string) {
-    const id = getId(rpcMethod, reqId);
-    const data = await this.store.get(id);
-    if (data == null) return false;
-    return true;
-  }
-
-  async isOverRate(
+  async limitApiCall(
     rpcMethod: string,
     reqId: string,
     offset: number = 1
-  ): Promise<boolean> {
+  ): Promise<
+    | {
+        current: number;
+        pttl: number;
+        isOverRate: boolean;
+      }
+    | {
+        isOverRate: false;
+      }
+  > {
     const id = getId(rpcMethod, reqId);
-    const data = await this.store.get(id);
-    if (data == null) return false;
-    if (this.rpcMethods[rpcMethod] == null) return false;
-
-    const count = +data + offset;
-    const maxNumber = this.rpcMethods[rpcMethod];
-    if (count > maxNumber) {
-      return true;
+
+    const maxNumber: number | undefined = this.rpcMethods[rpcMethod];
+
+    // No limit for this RPC
+    if (maxNumber == null) {
+      return {
+        isOverRate: false,
+      };
     }
-    return false;
+
+    const result: [string, string] = (await this.store.evalsha(
+      await this.getEvalSha(),
```

Comment:
The script may become unavailable on redis due to e.g. `SCRIPT FLUSH`. You would need to reload it then. https://redis.io/docs/manual/programmability/eval-intro/#cache-volatility

Reply:
![image](https://github.com/godwokenrises/godwoken/assets/13375784/48dfc289-bd8d-42c8-a1e8-a939b6edf774)

Worker process will exit and restart when cache invalidation, this request will failed and will recover soon.

Cache invalidation is a small probability event, so we may not have to deal with it specifically ?

Reply:
![image](https://github.com/godwokenrises/godwoken/assets/13375784/ff257573-5d11-47bc-bbaa-9a992b147451)

On second thought it would be better to handle the exception.